File: pal/lib/src/serialize.dart
import 'ast.dart';
import 'parse.dart';

String serializeProgram(Program p, {int lineLength = 80}) {
  String serializeBinding(Binding b) {
    String result = b.id;
    final type = b.type?.serializeExprIndent(lineLength, withFullHoleNames: false);
    final value = b.value?.serializeExprIndent(lineLength, withFullHoleNames: false);
    if (type != null) {
      result += ': ';
      var lines = type.split('\n');
      if (lines.first.length <= 100 - result.length) {
        result += lines.first;
        lines = lines.tail;
        if (lines.isNotEmpty) {
          result += '\n';
          result += lines.join('\n');
        }
      } else {
        result += '\n';
        result += lines.join('\n').indent;
      }
    }
    if (value != null) {
      result += ' = ';
      var lines = value.split('\n');
      if (lines.first.length <= 100 - result.split('\n').last.length) {
        result += lines.first;
        lines = lines.tail;
        if (lines.isNotEmpty) {
          result += '\n';
          result += lines.join('\n');
        }
      } else {
        result += '\n';
        result += lines.join('\n').indent;
      }
    }
    return result;
  }

  String serializeModule(List<Binding> m) => m.map(serializeBinding).join('\n\n');

// ignore: prefer_interpolation_to_compose_strings
  return p.map(serializeModule).join('\n\n--------------------\n\n') + '\n';
}

bool _withFullHoleNames = false;

extension Serialize on Expr {
  bool get isHole => switch (this) { Var(:var id) when id.startsWith('_') => true, _ => false };

  String _serializeApp(bool implicit, List<Expr> args) {
    switch (this) {
      case App(implicit: var im, :var fn, :var arg) when implicit == im:
        return fn._serializeApp(implicit, [arg, ...args]);
      default:
        return '${this._serializeExpr()}${args.map((arg) => arg._serializeExpr()).join(', ').parenthesize(implicit ? '<' : '(')}';
    }
  }

  String _serializeArg(FnKind kind, (ID?, Expr) arg) => switch ((kind, arg)) {
        (Fn.Typ, (null, var expr)) => expr._serializeExpr(),
        (Fn.Typ, (var id, var expr)) when expr.isHole =>
          '${id == null || id.startsWith('_') ? '_' : id}:',
        (Fn.Typ, (var id, var expr)) => '$id: ${expr._serializeExpr()}',
        (Fn.Def, (var id, var expr)) when expr.isHole =>
          id == null || id.startsWith('_') ? '_' : id,
        (Fn.Def, (null, var expr)) => ':${expr._serializeExpr()}',
        (Fn.Def, (var id, var expr)) =>
          '${id == null || id.startsWith('_') ? '_' : id}: ${expr._serializeExpr()}',
      };

  String _combineArgs(FnKind kind, List<(ID?, Expr)> args) =>
      args.map((arg) => _serializeArg(kind, arg)).join(', ');

  String _serializeFn(
    bool implicit,
    FnKind kind,
    List<(ID?, Expr)> args,
    List<(ID?, Expr)> explicitArgs,
  ) {
    switch (this) {
      case Fn(implicit: var im, kind: var thisKind, :var argID, :var argType, :var result)
          when thisKind == kind && !(implicit == false && im == true):
        if (im) {
          return result._serializeFn(im, kind, [...args, (argID, argType)], explicitArgs);
        } else {
          return result._serializeFn(im, kind, args, [...explicitArgs, (argID, argType)]);
        }

      default:
        final argPart = args.isEmpty ? '' : _combineArgs(kind, args).parenthesize('<');
        final explicitArgPart =
            explicitArgs.isEmpty ? '' : _combineArgs(kind, explicitArgs).parenthesize('(');
        final paren = kind == Fn.Def ? '{' : '[';
        final bodyPart = this is Var
            ? this._serializeExpr().parenthesize(paren)
            : ' ${' ${this._serializeExpr()} '.parenthesize(paren)}';

        return '$argPart$explicitArgPart$bodyPart';
    }
  }

  String _serializeExpr() {
    switch (this) {
      case Var(:var id) when id.startsWith('_') && !_withFullHoleNames:
        return '_';
      case Var(:var id):
        return id;
      case App(:var implicit):
        return this._serializeApp(implicit, []);
      case Fn(:var implicit, :var kind):
        return this._serializeFn(implicit, kind, [], []);
    }
  }

  String _serializeAppIndent(int colRemaining, bool implicit, List<Expr> args) {
    switch (this) {
      case App(implicit: var im, :var fn, :var arg) when implicit == im:
        return fn._serializeAppIndent(colRemaining, implicit, [arg, ...args]);
      default:
        final oneLine = _serializeApp(implicit, args);
        if (oneLine.length < colRemaining) {
          return oneLine;
        }
        final lines = args.map((arg) => arg._serializeExprIndent(colRemaining - 3).indent);
        final paren = implicit ? '<' : '(';
        return '''
${this._serializeExprIndent(colRemaining)}$paren
${lines.join(',\n')}
${MATCHING_PAREN[paren]}''';
    }
  }

  String _serializeFnIndent(
    int colRemaining,
    bool implicit,
    FnKind kind,
    List<(ID?, Expr)> args,
    List<(ID?, Expr)> explicitArgs,
  ) {
    switch (this) {
      case Fn(implicit: var im, kind: var thisKind, :var argID, :var argType, result: var body)
          when thisKind == kind && !(implicit == false && im == true):
        if (im) {
          return body._serializeFnIndent(
              colRemaining, im, kind, [...args, (argID, argType)], explicitArgs);
        } else {
          return body._serializeFnIndent(
              colRemaining, im, kind, args, [...explicitArgs, (argID, argType)]);
        }
      default:
        final oneLine = _serializeFn(implicit, kind, args, explicitArgs);
        if (oneLine.length < colRemaining) {
          return oneLine;
        }
        final argPart = args.isEmpty ? '' : _combineArgs(kind, args).parenthesize('<');
        final explicitArgPart =
            explicitArgs.isEmpty ? '' : _combineArgs(kind, explicitArgs).parenthesize('(');
        final paren = kind == Fn.Def ? '{' : '[';
        return '''$argPart$explicitArgPart $paren
${this._serializeExprIndent(colRemaining - 2).indent}
${MATCHING_PAREN[paren]}''';
    }
  }

  String _serializeExprIndent(int colRemaining) {
    switch (this) {
      case Var():
        return _serializeExpr();
      case App(:var implicit):
        return this._serializeAppIndent(colRemaining, implicit, []);
      case Fn(:var implicit, :var kind):
        return this._serializeFnIndent(colRemaining, implicit, kind, [], []);
    }
  }

  String serializeExprIndent(int colRemaining, {bool withFullHoleNames = true}) {
    _withFullHoleNames = withFullHoleNames;
    return _serializeExprIndent(colRemaining);
  }
}

extension StringOps on String {
  String get indent => splitMapJoin('\n', onNonMatch: (s) => '  $s');
  String parenthesize(String paren) => '$paren$this${MATCHING_PAREN[paren]}';
  String wrap(String ctx) => ctx.isEmpty ? this : '$ctx\n\n$this';
}

extension<T> on List<T> {
  List<T> get tail => sublist(1);
}

File: paljs/src/lang/serialize.ts
import {Binding, Expr, FnKind, ID, Program, Ctx, isApp, isFn, isHole, isVar} from './ast';
import {MATCHING_PAREN, parenFor} from './parse';

const parenthesize = (s: string, p: keyof typeof MATCHING_PAREN) => `${p}${s}${MATCHING_PAREN[p]}`;
export const indent = (s: string, times: number = 1) =>
  s
    .split('\n')
    .map((line) => (line.length === 0 ? line : `${'  '.repeat(times)}${line}`))
    .join('\n');

export const serializeBinding = (
  b: Binding,
  {lineLength = 80, withFullHoleNames}: {lineLength?: number; withFullHoleNames?: boolean} = {}
) => {
  let result = b.id;
  const type = b.type ? serializeExprIndent(lineLength, b.type, withFullHoleNames) : undefined;
  const value = b.value ? serializeExprIndent(lineLength, b.value, withFullHoleNames) : undefined;
  if (type != null) {
    result += ': ';
    let lines = type.split('\n');
    if (lines[0].length <= 100 - result.length) {
      result += lines[0];
      lines = lines.slice(1);
      if (lines.length !== 0) {
        result += '\n';
        result += lines.join('\n');
      }
    } else {
      result += '\n';
      result += indent(lines.join('\n'));
    }
  }
  if (value != null) {
    result += ' = ';
    let lines = value.split('\n');
    if (lines[0].length <= 100 - result.split('\n')[result.split('\n').length - 1].length) {
      result += lines[0];
      lines = lines.slice(1);
      if (lines.length !== 0) {
        result += '\n';
        result += lines.join('\n');
      }
    } else {
      result += '\n';
      result += indent(lines.join('\n'));
    }
  }
  return result;
};

const serializeModule = (
  m: Binding[],
  {
    lineLength = 80,
    lineSep = 2,
    withFullHoleNames,
  }: {lineLength: number; lineSep: number; withFullHoleNames?: boolean}
) => m.map((b) => serializeBinding(b, {lineLength, withFullHoleNames})).join('\n'.repeat(lineSep));

export const serializeCtx = (ctx: Ctx, lineLength = 80) => {
  return (
    '{\n' +
    indent(
      serializeModule(
        Object.entries(ctx).map(([id, {type, value}]) => ({id, type, value})),
        {lineLength: lineLength - 2, lineSep: 1, withFullHoleNames: true}
      )
    ) +
    '\n}\n'
  );
};

export const serializeProgram = (p: Program, {lineLength = 80}: {lineLength: number}) => {
  return (
    p
      .map((m) => serializeModule(m, {lineLength, lineSep: 2, withFullHoleNames: false}))
      .join('\n\n--------------------\n\n') + '\n'
  );
};

let _withFullHoleNames = false;

const serializeApp = (e: Expr, implicit: boolean, args: Expr[]): string => {
  if (isApp(e) && e.implicit === implicit) {
    return serializeApp(e.fn, implicit, [e.arg, ...args]);
  }
  const argString = parenthesize(args.map(serializeExpr).join(', '), parenFor(implicit));
  return `${serializeExpr(e)}${argString}`;
};

const serializeArg = (kind: FnKind, [argID, expr]: [ID | undefined, Expr]) => {
  const result = (() => {
    const idPart = isHole(argID) ? '_' : argID;
    const exprPart = serializeExpr(expr);
    if (kind === 'type') {
      if (idPart === '_') return exprPart;
      if (exprPart === '_') return `${idPart}:`;
      return `${idPart}: ${exprPart}`;
    } else {
      if (exprPart === '_') return idPart;
      if (idPart === '_') return `:${exprPart}`;
      return `${idPart}: ${exprPart}`;
    }
  })();
  return result;
};

const combineArgs = (implicit: boolean, kind: FnKind, args: [ID | undefined, Expr][]) =>
  args.length === 0
    ? ''
    : parenthesize(args.map((arg) => serializeArg(kind, arg)).join(', '), parenFor(implicit));

const serializeFn = (
  e: Expr,
  implicit: boolean,
  kind: FnKind,
  args: [ID | undefined, Expr][],
  explicitArgs: [ID | undefined, Expr][]
): string => {
  if (isFn(e) && e.kind === kind && !(implicit === false && e.implicit === true)) {
    if (e.implicit) {
      return serializeFn(e.result, e.implicit, kind, [...args, [e.argID, e.argType]], explicitArgs);
    } else {
      return serializeFn(e.result, e.implicit, kind, args, [...explicitArgs, [e.argID, e.argType]]);
    }
  }

  const argPart = combineArgs(true, kind, args);
  const explicitArgPart = combineArgs(false, kind, explicitArgs);
  const bodyPart = isVar(e)
    ? parenthesize(serializeExpr(e), parenFor(kind))
    : ' ' + parenthesize(` ${serializeExpr(e)} `, parenFor(kind));

  return `${argPart}${explicitArgPart}${bodyPart}`;
};

const serializeExpr = (e: Expr): string => {
  if (isVar(e)) {
    if (isHole(e) && !_withFullHoleNames) {
      return '_';
    }
    return e.id;
  } else if (isApp(e)) {
    return serializeApp(e, e.implicit, []);
  } else {
    return serializeFn(e, e.implicit, e.kind, [], []);
  }
};

const serializeAppIndent = (
  colRemaining: number,
  e: Expr,
  implicit: boolean,
  args: Expr[]
): string => {
  if (isApp(e) && e.implicit === implicit) {
    return serializeAppIndent(colRemaining, e.fn, implicit, [e.arg, ...args]);
  }
  const oneLine = serializeApp(e, implicit, args);
  if (oneLine.length < colRemaining) {
    return oneLine;
  }
  const lines = args.map((arg) => indent(_serializeExprIndent(colRemaining - 3, arg)));
  return `${_serializeExprIndent(colRemaining, e)}${parenFor(implicit)}
${lines.join(',\n')}
${MATCHING_PAREN[parenFor(implicit)]}`;
};

const serializeFnIndent = (
  colRemaining: number,
  e: Expr,
  implicit: boolean,
  kind: FnKind,
  args: [ID | undefined, Expr][],
  explicitArgs: [ID | undefined, Expr][]
): string => {
  if (isFn(e) && e.kind === kind && !(implicit === false && e.implicit === true)) {
    if (e.implicit) {
      return serializeFnIndent(
        colRemaining,
        e.result,
        e.implicit,
        kind,
        [...args, [e.argID, e.argType]],
        explicitArgs
      );
    } else {
      return serializeFnIndent(colRemaining, e.result, e.implicit, kind, args, [
        ...explicitArgs,
        [e.argID, e.argType],
      ]);
    }
  }

  const oneLine = serializeFn(e, implicit, kind, args, explicitArgs);
  if (oneLine.length < colRemaining) {
    return oneLine;
  }
  const argPart = combineArgs(true, kind, args);
  const explicitArgPart = combineArgs(false, kind, explicitArgs);

  return `${argPart}${explicitArgPart} ${parenFor(kind)}
${indent(_serializeExprIndent(colRemaining - 2, e))}
${MATCHING_PAREN[parenFor(kind)]}`;
};

const _serializeExprIndent = (colRemaining: number, e: Expr): string => {
  if (isVar(e)) return serializeExpr(e);
  if (isApp(e)) return serializeAppIndent(colRemaining, e, e.implicit, []);
  return serializeFnIndent(colRemaining, e, e.implicit, e.kind, [], []);
};

export const serializeExprIndent = (
  colRemaining: number,
  e: Expr,
  withFullHoleNames: boolean = true
) => {
  _withFullHoleNames = withFullHoleNames;
  return _serializeExprIndent(colRemaining, e);
};
